Blueprint.Api/Services/InjectService.cs

```diff
@@ -0,0 +1,231 @@
+                .Where(i => i.CatalogId == catalogId)
+                .Select(m => m.Inject)
+                .ToListAsync(ct);
+            return _mapper.Map<IEnumerable<Inject>>(injects);
+        }
+
+        public async Task<ViewModels.Inject> GetAsync(Guid id, CancellationToken ct)
+        {
+            var item = await _context.Injects
+                .Include(se => se.DataValues)
+                .SingleAsync(a => a.Id == id, ct);
+
+            if (item == null)
+                throw new EntityNotFoundException<InjectEntity>();
+
+            if (!(await _authorizationService.AuthorizeAsync(_user, null, new ContentDeveloperRequirement())).Succeeded)
+            {
+                var userId = _user.GetId();
+                var unitIdList = await _context.CatalogUnits
+                    .Where(m => m.Id == userId)
+                    .Select(m => m.UnitId)
+                    .ToListAsync(ct);
+                var catalogIdList = await _context.CatalogUnits
+                    .Where(m => unitIdList.Contains(m.UnitId))
+                    .Select(m => m.CatalogId)
+                    .ToListAsync(ct);
+                var isAuthorized = await _context.CatalogInjects
+                    .Where(m => catalogIdList.Contains(m.CatalogId))
+                    .AnyAsync(ct);
+                if (!isAuthorized)
+                    throw new ForbiddenException(); 
+            }
+
+            return _mapper.Map<ViewModels.Inject>(item);
+        }
+
+        public async Task<ViewModels.Inject> CreateAsync(Guid catalogId, Inject inject, CancellationToken ct)
+        {
+            // user must be a Content Developer
+            if (!(await _authorizationService.AuthorizeAsync(_user, null, new ContentDeveloperRequirement())).Succeeded)
+                throw new ForbiddenException();
+
+            // start a transaction, because we may also update DataValues
+            await _context.Database.BeginTransactionAsync();
+            // create the inject
+            inject.Id = inject.Id != Guid.Empty ? inject.Id : Guid.NewGuid();
+            inject.DateCreated = DateTime.UtcNow;
+            inject.CreatedBy = _user.GetId();
+            inject.DateModified = null;
+            inject.ModifiedBy = null;
+            var injectEntity = _mapper.Map<InjectEntity>(inject);
+            _context.Injects.Add(injectEntity);
+            await _context.SaveChangesAsync(ct);
+            // create the associated data values
+            var dataFieldList = await _context.InjectTypes
+                .Where(m => m.Id == inject.InjectTypeId)
+                .Select(m => m.DataFields)
+                .SingleOrDefaultAsync(ct);
+            foreach (var dataField in dataFieldList)
+            {
+                var dataValue = inject.DataValues
+                    .FirstOrDefault(dv => dv.DataFieldId == dataField.Id);
+                if (dataValue == null)
+                {
+                    dataValue = new DataValue();
```

Comment:
Here and in the UpdateAsync do we need to check if the new DataValue has a DataField that is not part of the "schema" of the Inject? If it does do we add that to the schema to keep it in sync? Disallow the add or update?

Reply:
We just allow it.  It is possible that the user could delete a dataField and then later on, add it back in.  This way, they don't lose any of the previously entered data.  We only display the dataValues for dataFields defined in the InjectType.  This just keeps the most flexibility.  An inject, in theory, could be included in catalogs that have different sets of dataFields.  We can allow or disallow that, but if data has been added to an inject, I think it is good to maintain it.  This is one of the trade-offs for allowing dynamic dataField creation.  I think it is acceptable to not have strict data integrity in this case.